Source/DafnyCore/Resolver/NameResolutionAndTypeInference.cs

```diff
@@ -21,6 +21,120 @@
 
 namespace Microsoft.Dafny {
   public partial class Resolver {
+    /// <summary>
+    /// There are two rounds of name resolution + type inference. The "initialRound" parameter says which one to do.
+    /// </summary>
+    void ResolveNamesAndInferTypes(List<TopLevelDecl> declarations, bool initialRound) {
+      foreach (TopLevelDecl topd in declarations) {
+        Contract.Assert(topd != null);
+        Contract.Assert(VisibleInScope(topd));
+        Contract.Assert(AllTypeConstraints.Count == 0);
+        Contract.Assert(currentClass == null);
+
+        ResolveNamesAndInferTypesForOneDeclaration(topd, initialRound);
+
+        Contract.Assert(AllTypeConstraints.Count == 0);
+        Contract.Assert(currentClass == null);
+      }
+    }
+
+    void ResolveNamesAndInferTypesForOneDeclaration(TopLevelDecl topd, bool initialRound) {
```

Comment:
Seems like two methods would be better than one method with the `initialRound` parameter. The shared code, which seems to be very littles, can be put into methods.